### ISSUE TITLE
How do pre-authorized transactions actually work?

### ISSUE THREAD

**Issue by vladfaust** [stellar/stellar-protocol#73]
In my opinion, [current documentation](https://www.stellar.org/developers/guides/concepts/multi-sig.html#pre-authorized-transaction) lacks of good example.

I came up with a simple "escrow" model. Let's say A wants to sell his 30 XLM offchain for real-world USD and B wants to buy it. An escrow account C is set up with two pre-authorized transactions which will result either in successful transfer or cancellation. Therefore, only four outcomes are possible:

- B sends USD and gets his 30 XLM, A is returned with 20 XLM (success)
- Both A and B agree on cancellation and A gets his 50 XLM back (cancel)
- B sends USD but A goes away losing his 50 XLM (A frauds and gets punished)
- B neither sends USD not cancels, so A is fu**ed up (the worst scenario)

Let's skip postponed refunding or B's insurance deposit for the sake of simplicity.

```yaml
# tx1
source: A
seq: A's sequence
ops:
  - create escrow account C with 50 XLM balance
  - add tx2 hash as signer with weight 1
  - add tx3 hash as signer with weight 1
  - set master key weight to 0
signers:
  - A
  - C

# tx2 (success)
source: ?
seq: C's sequence
ops:
  - send 30 XLM to B
  - merge C with A
signers:
  - ?

# tx3 (cancel)
source: ?
seq: C's sequence
ops:
  - send 30 XLM to A
  - merge C with A
signers:
  - ?
```

As far as I understand, to initiate the exchange, A has to sign tx1 and what's next? Who signs tx2 and tx3 and how will it work out? If we try to submit tx2 or tx3 signed by C it wont work out because C's master weight is already set to 0...

P.S: This paragraph is messed up:

> This type of signer is especially usefull in escrow accounts. You can create a pre-authorized transaction for two different transactions each with the same sequence number but that send to two different accounts. This means that only one of the pre-authorized transactions or the other will be able to execute.

You call "pre-authorized transaction" tx1 in the first sentence but in the last tx2 and tx3 become pre-authorized.

**Comment by vogel** [comment]
You do not need to sign tx2 and tx3, but their source must be set to C.
C already accepted both of them. If you set both tx2 and tx3 sequence number to the same value, you will only be able to execute one of them. In reality, anyone will be able to execute them at any moment, so it is not the best implementation of escrow.

And you are right, the paragraph in documentation is messes up. I'll try to fix that.

**Comment by vladfaust** [comment]
Is it possible to require A to sign tx2 but require B to sign tx3?

**Comment by vogel** [comment]
You can add no-ops ops to tx2 and tx3. For example:

tx2 ops:
- A pays A 0.00001 XLM
- .....

tx3 ops:
- B pays B 0.00001 XLM
- .....

**Comment by vogel** [comment]
It will work as long as B exists when the transaction is executed.

**Comment by vladfaust** [comment]
@vogel, thanks for such fast responses :heart:, however please let me ask a few more questions to make things absolutely clear.

What if B was not created yet in this case? I mean, if user B had only had a keypair, not the funded account itself, and tx2 would have included `Create Account B` operation, and user A could not execute tx3 without B's signature - is that somehow possible with current Stellar state?

**Comment by vogel** [comment]
It would mean that tx2 would have to be executed before tx3, which is only possible if you create them with increasing sequence numbers.

**Comment by vladfaust** [comment]
@vogel you know what, I've just tried to reproduce the initial example and stuck with an issue of C's sequence number. I literally can not obtain C's sequence in neither tx2 or tx3 because it's not created yet, so I cant even build build them to pre-authorize!

**Comment by vogel** [comment]
Is it resolved then, can we close that issue?

**Comment by vladfaust** [comment]
I think it still worth updating documentation...

**Comment by vogel** [comment]
Is that OK? https://github.com/stellar/docs/commit/5f650ee1c7258f27e0d8a4d1a76664009fa9576a

**Comment by vladfaust** [comment]
**I'm not excellent in English, but this piece text could describe it more clearly in my opinion:**

It is possible for an account to pre-authorize a particular transaction by adding its hash as a signer to the account. To do that you have to prepare a transaction beforehand with a proper sequence number. Then you can calculate the hash of this transaction and add it as a signer to the account. When such a pre-authorized transaction is submitted to the network, it is already been signed by the account, therefore no additional signing is required at that moment.

Signers of this type are automatically removed from an account when a matching transaction is properly applied. In case of error, or when the matching transaction is never submitted, the signer remains and must be manually removed using the SetOptions operation.

This type of signers is useful for escrow accounts. For example, you can create an escrow account with two different pre-authorized transactions as its signers, each with the same sequence number but sending a payment either to A or to B. As a result, only one of the these pre-authorized transactions could be applied in the future, meaning that the payment could only be sent either to A or to B.

**Comment by cryptomental** [comment]
Your description allowed me to, at least partially, understand this feature (thumbsup).

I would like to implement a variation of @vladfaust 's with a simple time-limited escrow operation but without creating account C every time. Trusted escrow account could always stay the same. 

The idea is to try out Stellar for e.g. online shops use case with a guarantee that the product that was sent works as expected and buyer can try it out for some time, let's say 1 day after it arrived. During that time the buyer can ask for a refund, no strings attached and funds would be automatically sent back to him.

A is the buyer, B is the seller. A sends to B via C. C is the escrow account. 
Can you help me on how to design this flow?
 
```yaml
# tx1 - send A to C and add tx2 and tx3 transaction signatures to escrow account
source: A
seq: A's sequence
ops:
  - send A to C (deposit in escrow account)
  - add tx2 hash as C signer with weight 1
  - add tx3 hash as C signer with weight 1
signers:
  - A
  - C

# tx2 - send funds to B after 1 day
source: C
seq: C's sequence X
ops:
  - can be triggered 24h after tx1
  - send C to B
signers:
  - A & C or C only?

# tx3 - ask for a refund
source: C
seq: C's sequence X
ops:
  - time-bound max 24h
  - send C to A
signers:
  - A & C
```


I am still a bit confused with weights, signers and the sequence and how to set it up properly. 

My understanding is that tx2 and tx3 must have the same sequence number. Should tx1 be submitted to the network right after the purchase, tx3 can be submitted within 24 hours and tx2 can be submitted only after 24 hours or also earlier and tx3 would invalidate it?

**Comment by vogel** [comment]
@vladfaust could you make PR with that change?

**Comment by ckmakcesar** [comment]
Still kind of confused how it works. Would anyone help on the following example? 

> 2 accounts: Alice & Bob
> 
> Last used seq#: Alice = 500
> Last used seq#: Bob  = 600
> 
> <step 1>
> Alice pays Bob 20XLM
> seq# = 501
> tx_hash -> "XXAAAA"
> not signed yet
> 
> <step 2>
> Bob adds tx_hash "XXAAAA" as signer (weight=1)
> seq# = 601
> signed with Bob's SecretKey
> > Successful
> 
> <step 3>
> Alice pays Bob 20XLM
> seq# = 501
> signed with Alice's SecretKey
> > Fail 400: tx_bad_auth_extra

Anywhere I should correct my concept?

**Comment by vogel** [comment]
Please try without additional signature on step3:

```
<step 3>
Alice pays Bob 20XLM
seq# = 501
not extra-signed as it is pre-signed
```

**Comment by ckmakcesar** [comment]
I was trying with Laboratory and it requires a signature anyway. Maybe I should try again with SDK. 
![image](https://user-images.githubusercontent.com/31118333/38302533-9a508f24-3835-11e8-84f3-d1dbe189bc5d.png)

**Comment by vogel** [comment]
Use this endpoint: https://www.stellar.org/laboratory/#explorer?resource=transactions&endpoint=create&network=test

**Comment by ckmakcesar** [comment]
Thanks Rafał @vogel !

**Comment by jedmccaleb** [comment]
More appropriate for stackoverflow: https://stellar.stackexchange.com/
